{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "22a662a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ACGT': 27,\n",
       " 'CGTT': 27,\n",
       " 'GTTG': 27,\n",
       " 'TTGC': 27,\n",
       " 'TGCA': 27,\n",
       " 'GCAT': 27,\n",
       " 'CATG': 27,\n",
       " 'ATGT': 27,\n",
       " 'TGTC': 27,\n",
       " 'GTCG': 27,\n",
       " 'TCGC': 27,\n",
       " 'CGCA': 27,\n",
       " 'ATGA': 27,\n",
       " 'TGAT': 27,\n",
       " 'GATG': 27,\n",
       " 'ATGC': 27,\n",
       " 'TGAG': 27,\n",
       " 'GAGA': 27,\n",
       " 'AGAG': 27,\n",
       " 'GAGC': 27,\n",
       " 'AGCT': 27}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def hammingdistance(p,q):\n",
    "    distance = 0\n",
    "    for i in range(len(p)):\n",
    "        if p[i] != q[i]:\n",
    "            distance += 1\n",
    "    return distance\n",
    "def Approximatepatterncount(Text,pattern,d):\n",
    "    count = 0\n",
    "    for i in range(len(Text)-len(pattern)+1):\n",
    "        if hammingdistance(Text[i:i+1],pattern) <= d:\n",
    "            count += 1\n",
    "    return count\n",
    "def freqmap(Text,k,d):\n",
    "    map = {}\n",
    "    for i in range(len(Text)-k+1):\n",
    "        pattern = Text[i:i+k]\n",
    "        map[pattern] = Approximatepatterncount(Text,pattern,d)\n",
    "    return map\n",
    "def frequentword(Text,k,d):\n",
    "    words = []\n",
    "    freq = freqmap(Text,k,d)\n",
    "    m = max(freq.values())\n",
    "    for pattern in freq:\n",
    "        if freq[pattern] == m:\n",
    "            words.append(pattern)\n",
    "    return words\n",
    "\n",
    "freqmap(\"ACGTTGCATGTCGCATGATGCATGAGAGCT\",4,1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6666a683",
   "metadata": {},
   "source": [
    "## good try but this is'nt the actual thing asked"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbf6e969",
   "metadata": {},
   "source": [
    "# Frequency with mismatch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "557f3f53",
   "metadata": {},
   "source": [
    "## Plain English\n",
    "\n",
    "- Make an empty frequency map freqMap.\n",
    "\n",
    "- For every k-mer substring in Text:\n",
    "\n",
    "- Generate its d-neighborhood (all k-mers within d mismatches).\n",
    "\n",
    "- Count each neighborâ€™s occurrences in freqMap.\n",
    "\n",
    "- Find the maximum frequency in the map.\n",
    "\n",
    "- Collect all patterns with that maximum count.\n",
    "\n",
    "- Return them."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea018219",
   "metadata": {},
   "source": [
    "## personal thought\n",
    "\n",
    "After learning about the neighbour algorithm;\n",
    "\n",
    "I thought what I could do is simply traditionally appling the ``frequencymap(genome,k)`` algorithm ; followed by the `mostfrequentword(genome,k)` and then applying the `neighbour(pattern,d)` algorithm to the *mostfrequentword* or *the most frequent pattern* recived;\n",
    "\n",
    "But this is'nt the right approach and doing this we are obyiously  getting the neighbourhoods of the most frequent pattern ; but are those pattern present in the map framed or are those actually even present in the text or genome  peovided ??????? \n",
    "\n",
    "### !!!! NO !!!!\n",
    "\n",
    "### And here I was Wrong !!!!\n",
    "\n",
    "So instead start with the EMPTY DICTIONARY and then fix a pattern and apply NEIGHBOU algo. to it and then one by one check if its present in that map or not and then keep on iteraing the count from `+=1` with occurance of that neighbour every time it comes \n",
    "\n",
    "And hence u got the reqired algo. for `FrequencyMapWithMismatch`\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f8ab22b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATGT', 'GATG', 'ATGC']\n",
      "AAAAAAA\n"
     ]
    }
   ],
   "source": [
    "def HammingDistance(p,q):\n",
    "    distance = 0\n",
    "    for i in range(len(p)):\n",
    "        if p[i] != q[i]:\n",
    "            distance += 1\n",
    "    return distance\n",
    "\n",
    "def Neighbour(pattern,d):\n",
    "    nucleotides = [\"A\",\"T\",\"G\",\"C\"]\n",
    "\n",
    "    if d == 0:\n",
    "        return{pattern}\n",
    "    if len(pattern) == 1:\n",
    "        return set(nucleotides)\n",
    "    \n",
    "    suffix = pattern[1:]\n",
    "    suffixneighbour = Neighbour(suffix,d)\n",
    "    neighbourhood = set()\n",
    "    for text in suffixneighbour:\n",
    "        if HammingDistance(text,suffix) < d:\n",
    "            for x in nucleotides:\n",
    "                neighbourhood.add(x + text)\n",
    "        else:\n",
    "            neighbourhood.add(pattern[0] + text)    \n",
    "    \n",
    "    return neighbourhood\n",
    "\n",
    "def FreqMapWithMismatch(genome,k,d):\n",
    "    freqmap = {}\n",
    "    n = len(genome)\n",
    "    for i in range(len(genome)-k+1):\n",
    "        pattern = genome[i:i+k]\n",
    "        Neighbourhood = Neighbour(pattern,d)\n",
    "        for neighbour in Neighbourhood:\n",
    "            if neighbour not in freqmap:\n",
    "                freqmap[neighbour] = 1\n",
    "            else:\n",
    "                freqmap[neighbour] += 1\n",
    "\n",
    "    m = max(freqmap.values())\n",
    "    patterns = []\n",
    "    for pattern in freqmap:\n",
    "        if freqmap[pattern] == m:\n",
    "            patterns.append(pattern)\n",
    "    return patterns\n",
    "\n",
    "print(FreqMapWithMismatch(\"ACGTTGCATGTCGCATGATGCATGAGAGCT\", 4, 1))\n",
    "\n",
    "print(*FreqMapWithMismatch(\"TGATAATATGATAAAACCAAAAAATACCCAATACCCAAAAAAAAAAAAAAAATAAATAAATAAATAAAAAAATAAAAATACAAAATACAAAACCAATACCCAAAATATACAATAAATAAATAAAAAAATAAAAACCAATACTGATAAAAAAAAAAAATACTGATAATAAATAAAAATGATTACAAAAAATATGATCCAACCAAAATAAATATGATTACAATAAATAAATACCAAAAAATACTACTACAATATGATAATAAAAATGATAATAAAAAAAAATACAAAAAAAATACAAAAAAAATGATCCAAAATATGATAATAAATACCAATACTACCCAATGATAATAAAAAAATAAAAACCAAAAAAAAAAAATATGATTACAATAAAAATAC\",7,2),sep = \" \")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b55d51f0",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b99bd06b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ATA', 'ACA', 'AAC', 'TAA', 'GAA', 'AGA', 'CAA', 'AAG', 'AAT'}\n"
     ]
    }
   ],
   "source": [
    "# immediate neighbour (just for rivision) \n",
    "# we use this algo. is we want neighbour just with the difference of one algo.\n",
    "\n",
    "def immediateneighbour(pattern):\n",
    "    nucleotides = [\"A\",\"T\",\"G\",\"C\"]\n",
    "    n = len(pattern)\n",
    "    neignbourhood = set()\n",
    "    for i in range(n):\n",
    "        symbol = pattern[i]\n",
    "        for x in nucleotides:\n",
    "            if x != symbol:\n",
    "                neighbour = pattern[:i] + x + pattern[i+1:]\n",
    "                neignbourhood.add(neighbour)\n",
    "    return neignbourhood\n",
    "\n",
    "print(immediateneighbour(\"AAA\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a30cc11",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09f6002a",
   "metadata": {},
   "source": [
    "## Frequent words with mismatches and reverse complement problem"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35d98447",
   "metadata": {},
   "source": [
    "Now this is same as the previous one `(FreqMapWithMismatch)` \n",
    "\n",
    "but here comes the extra part too , after getting the most probable k-mer with given number of mismatches , make a variable to reverse the genome provided to pu in the `freqmapwithmismatches` , store it in the variable and then run the same algo. for the reversed genome and also keep the same dictionary and add the reversed most probable k-mer in that same dictionary as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "82aa8d49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43\n"
     ]
    }
   ],
   "source": [
    "def HammingDistance(p,q):\n",
    "    distance = 0\n",
    "    for i in range(len(p)):\n",
    "        if p[i] != q[i]:\n",
    "            distance += 1\n",
    "    return distance\n",
    "\n",
    "def Neighbour(pattern,d):\n",
    "    nucleotide = [\"A\",\"T\",\"G\",\"C\"]\n",
    "\n",
    "    if d == 0:\n",
    "        return {pattern}\n",
    "    if len(pattern) == 1:\n",
    "        return set(nucleotide)\n",
    "    \n",
    "    neighbourhood = set()\n",
    "    suffix = pattern[1:]\n",
    "    suffixneighbour = Neighbour(suffix,d)\n",
    "    for text in suffixneighbour:\n",
    "        if HammingDistance(text,suffix) < d:\n",
    "            for x in nucleotide:\n",
    "                neighbourhood.add(x + text)\n",
    "        else:\n",
    "            neighbourhood.add(pattern[0] + text)\n",
    "\n",
    "    return neighbourhood \n",
    "\n",
    "def ReverseComplement(pattern):\n",
    "    pattern = Reverse(pattern)\n",
    "    pattern = Complement(pattern)\n",
    "    return pattern\n",
    "\n",
    "def Reverse(pattern):\n",
    "    pattern = pattern[::-1]\n",
    "    return pattern\n",
    "\n",
    "def Complement(pattern):\n",
    "    comp = \"\"\n",
    "    dict = {\n",
    "        \"A\" : \"T\",\n",
    "        \"T\" : \"A\",\n",
    "        \"G\" : \"C\",\n",
    "        \"C\" : \"G\"\n",
    "    }\n",
    "    for char in pattern:\n",
    "        comp = comp + dict[char]\n",
    "    return comp\n",
    "\n",
    "def FreqWordsWithMismatchAndReverseComplement(genome,k,d):\n",
    "    freqmap = {}\n",
    "    n = len(genome)\n",
    "    for text in [genome,ReverseComplement(genome)]:\n",
    "        for i in range(len(text)-k+1):\n",
    "            pattern = text[i:i+k]\n",
    "            neighbourhood = Neighbour(pattern,d)\n",
    "            for neighbour in neighbourhood:\n",
    "                if neighbour not in freqmap:\n",
    "                    freqmap[neighbour] = 1\n",
    "                else:\n",
    "                    freqmap[neighbour] += 1\n",
    "\n",
    "    m = max(freqmap.values())\n",
    "    patterns = []\n",
    "    for pattern in freqmap:\n",
    "        if freqmap[pattern] == m:\n",
    "            patterns.append(pattern)\n",
    "    return patterns\n",
    "\n",
    "#print(*FreqWordsWithMismatchAndReverseComplement(\"TCGATTACGGATTTTTCGATTACTACTACGGTACTTTGGTCGTTTTACTACATTACTCGTTTTTTTTTATTCGATGGGGTCGTTTTTTATTACTTTTCGTCGTTTGGTTTTACTTTTACATTTTTCGATGGTACTTTATATGGTACTCGATTTTATTCGGGGGTACTCGATGGATTACGGTTTGGTCGATTTTGGTTTTCGATTACTCGGGTTTTTTGGATTTTTCGTTTATATTCGTACTCG\",6,3),sep = \" \")\n",
    "#print(FreqWordsWithMismatchAndReverseComplement(\"ACGTTGCATGTCGCATGATGCATGAGAGCT\",4,1))\n",
    "print(HammingDistance(\"CTTGAAGTGGACCTCTAGTTCCTCTACAAAGAACAGGTTGACCTGTCGCGAAG\",\"ATGCCTTACCTAGATGCAATGACGGACGTATTCCTTTTGCCTCAACGGCTCCT\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
